### ISSUE TITLE
Performance metrics in Autosubmit/GUI

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#524]
In GitLab by @mcastril on Apr 27, 2020, 12:47

Following the conversation started in https://earth.bsc.es/gitlab/es/autosubmit/issues/522#note_77443

> Yes, sometimes I have been looking in the GUI and missed this information, have real timestamps. Another twist, but very reliable with the information that we have, would be to filter all the *SIM* jobs (if there's any), get the total simulated time (sum of SIM running times avoiding job repetitions) and divide by the simulated time chunk*chunk_size*nchunks - startdate. That would give you an estimated number of SYPD. If you count queuing times you would have ASYPD. Well, we can define this a little bit more but this is something that definitely we will have one day. 

We will include in this issue the metrics that we can implement in Autosubmit/GUI and dicuss about their development.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 27, 2020, 12:58

@wuruchi , at this point @gmontanepinto , @orioljorba and the people from AC are interested in following changes in simulation times for the different testing suite test iterations.

In the GUI we currently have some tools to follow that information:

* The tree view provides Queuing/Running time per job. It is very visual and easy to see big changes in the simulation time for a given chunk.

![Captura_de_pantalla_de_2020-04-27_12-40-57](https://earth.bsc.es/gitlab/es/autosubmit/uploads/a363a7a2a8975c2a19806fb7380b3960/Captura_de_pantalla_de_2020-04-27_12-40-57.png)

* The Statistics section provides aggregated times per job/chunk, in a bar plot view with different series: Queued Run Failed Jobs_Failed Queued_Fail Run. And a summary report at the bottom.

![Captura_de_pantalla_de_2020-04-27_12-40-03](https://earth.bsc.es/gitlab/es/autosubmit/uploads/c6bf585c74b08315782814190ddff60e/Captura_de_pantalla_de_2020-04-27_12-40-03.png)

* Finally, there's another feature in the experiment's search view, allowing to see average running/queuing times for the jobs in an experiment (this is very useful to see failed jobs at a glance):

![Captura_de_pantalla_de_2020-04-27_12-52-51](https://earth.bsc.es/gitlab/es/autosubmit/uploads/52c9f32a1221d4f0389524f6b3b7734e/Captura_de_pantalla_de_2020-04-27_12-52-51.png)

So I think we have enough tools to work, as soon as they give reliable information. But I think there are two things with a lot of potential:

* @wuruchi do you think it would be easy to add another line in the Experiment summary (in the experiment's view) with the SIM average running time and SIM average queuing time, in case the experiment has SIM jobs? Both Auto-EC-Earth and Auto-MONARCH use the SIM nomenclature for their simulations, so having this we have much of the share.

* @wuruchi as a second step, do you think that this new data could be available in the Autosubmit API? Then @gmontanepinto and @pablohe could consume this info and serve it in the testing suite software. 

With these you would have information of the current testing suite in order to handle it as you wish. Afterwards we could discuss about comparing with previous experiments. For this we would need to access the information from previous runs.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 27, 2020, 12:59

I add @ macosta to the loop.

**Comment by kinow** [comment]
In GitLab by @ macosta on Apr 27, 2020, 13:05

I am happy that this is on going!

As you know CPMIP are standard performance metrics and our goal should be to include all of them through Autosubmit, we could have a telco for discussion

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 27, 2020, 13:20

Hello @mcastril 

Thanks for the detailed description of your requirements.

It should be possible to implement this in a reasonable time. I will update this issue as soon as I have something to show.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 27, 2020, 17:40

Hello @mcastril 

I have implemented the first version of this feature in: [https://earth.bsc.es/autosubmitapp/](https://earth.bsc.es/autosubmitapp/). Click on `Show Detailed Data` or `Summary` of any experiment. Then you will see in the summary the line:

SIM (`X`): avg. queue `Y` min. (`Z`) | run `A` min. (`B`)

where:

* `X`: Number of SIM jobs in the experiment.
* `Y`: Average Queuing time for a number `Z` of jobs with status `FAILED`, `SUBMITTED`, `QUEUING`, `RUNNING`, `COMPLETED`.
* `A`: Average Running time for a number `B` of jobs with status `FAILED`, `RUNNING`, `COMPLETED`.

The API request has been implemented, more information in: [https://earth.bsc.es/gitlab/wuruchi/autosubmitreact/-/wikis/Experiment-Summary](https://earth.bsc.es/gitlab/wuruchi/autosubmitreact/-/wikis/Experiment-Summary)

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 27, 2020, 18:10

Hi @wuruchi , thanks for your promptness. I think it is a very nice feature. @gmontanepinto , @pablohe , you have it ready in the API, so you can improve the testing suite to give this number in the report.

One thing that we have to think how to handle is the queuing times in experiments with wrappers. Jobs waiting in a running wrapper count as queuing, even if they are still not ready to run. This inflates the total and average queuing times for experiments with wrappers. We will look for a solution, this is only for you to take into account this fact when reading the numbers.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 28, 2020, 10:58

I think you can create an extra tab "Performance". The Tree View loads automatically when entering in the Experiment Page, so the information should be there I guess. What you could do is an automatic Refresh of the Tree View info when loading the Performance tab.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 29, 2020, 14:25

Hello @gmontanepinto 
 
> It is possible to show this time with more precision? At least for the case where it is less than 1 minute.

Sure, this can be fixed. I will comment here as soon as it is implemented.

> Is the summary based only on the last run of the experiment? If yes, would it be possible to get also the data from previous runs? Not by default, but maybe with an extra parameter in the API request.

I think this is possible since Autosubmit stores in the logs (_TOTAL_STATS files) the stats for all the runs of the corresponding jobs; however, I am not sure how reliable this could be. Perhaps @mcastril has something to say about this.

In case we try to implement this feature, it will take some time; there are some fundamental things that need to be changed in the way we are currently mapping information from Autosubmit logs to the central database that the API queries.

**Comment by kinow** [comment]
In GitLab by @gmontanepinto on Apr 29, 2020, 17:04

OK, for now we can just make the testing suite give the numbers of the current run and do a manual comparison. Thanks!

**Comment by kinow** [comment]
In GitLab by @ macosta on Apr 29, 2020, 17:11

Taking into account that some of the metrics that you are talking are closed to the CPMIP metrics and the performance job using TOTAL-STATS to calculate them. Does it make sense to think about the partial CPMIP implementation now? Could we do a telco next week maybe?

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 29, 2020, 17:14

Yes, we can start talking about it. This is something that we had in mind so it fits perfectly with our plans.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 29, 2020, 17:20

Scheduled time is Ok with me.

I am going to take a look at the paper.

**Comment by kinow** [comment]
In GitLab by @gmontanepinto on Apr 29, 2020, 18:13

mentioned in issue testing_suite#13

**Comment by kinow** [comment]
In GitLab by @wuruchi on Apr 29, 2020, 18:49

Hello @gmontanepinto @mcastril 

I have updated the GUI with the requested changes.

* Times are now shown in `datetime` format.
* Times are shown in bold for better visualization.
* API documentation has been updated: [https://earth.bsc.es/gitlab/wuruchi/autosubmitreact/-/wikis/Experiment-Summary](https://earth.bsc.es/gitlab/wuruchi/autosubmitreact/-/wikis/Experiment-Summary)

I wonder if I should also apply this format to the times shown in the Tree View.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Apr 30, 2020, 09:26

Okay, I'll take a look at the paper next week

**Comment by kinow** [comment]
In GitLab by @ macosta on Apr 30, 2020, 09:37

As Miguel was commenting, here you can find a summary only with the metrics:https://docs.google.com/document/d/12yWDwXsohf4G4MPeP6e3Eil4ZL-YeIN71dBcoWRliEg/edit

Probably take a look to the performance job of auto-ecearth could be also interesting because at the end we want to port those metrics already coded from this shell script to your autosubmit solution. You do not need to start from scratch ;)

**Comment by kinow** [comment]
In GitLab by @mcastril on May 8, 2020, 13:44

> I wonder if I should also apply this format to the times shown in the Tree View. 

Maybe yes. For very long jobs the number of minutes is not so direct.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 8, 2020, 19:03

Hi all,

As we decided in the meeting, Autosubmit should directly compute the following metrics:

```
Model
Platform
SYPD
ASYPD
RSYMD
CHSPY
```

As @ macosta said, the calculations can be looked in the performance metrics template in Auto-EC-Earth (to understand how they work).

In the Performance Metrics view in Autosubmit and Autosubmit GUI, another column for the ?YPD metrics translation to ?DPD (days per day) should be provided. 

For the rest of the metrics, we have to figure out a mechanism to communicate their values from the Auto-Models to Autosubmit.

Finally, we also have to figure out how to store different runs of the same experiment and distinguish between different instances (create an internal ID after a new create?)

**Comment by kinow** [comment]
In GitLab by @pablohe on May 13, 2020, 09:52

mentioned in issue auto-ecearth3#990

**Comment by kinow** [comment]
In GitLab by @mcastril on May 29, 2020, 15:16

@wuruchi has announced the availability of some metrics trough an API:

https://earth.bsc.es/gitlab/es/testing_suite/issues/13#note_83930

I checked one experiment that I launched recently:

http://192.168.11.91:8081/performance/a2vp

* The SYPD is very similar to the one provided by the EC-Earth runscript: 4.212 (API) vs 4.24 (script). This may be because Autosubmit counts some more seconds for the job to be terminated. This was a very short job, so the difference should be higher that in real simulations, and in any case it was only 1%.

* Same for CHSY: 7111.17 against 7303. In this case it's a 2% difference, but we should compare using a longer experiment. 

* No comparison for the ASYPD, but it is reasonable taking into account that it was 3 hour queuing to only run for 10 minutes.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 29, 2020, 15:19

@wuruchi , I am looking at t0e0 but I can only see up to chunk 89 in the API, while we have logs until chunk 90.

**Comment by kinow** [comment]
In GitLab by @wuruchi on May 29, 2020, 15:25

For some reason, chunk 90 appears near the end of the answer, but it is there.

http://192.168.11.91:8081/performance/t0e0

I will try to improve the order.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 4, 2020, 11:13

As @wuruchi announced in https://earth.bsc.es/gitlab/es/testing_suite/issues/13#note_84522

There is a first implementation of the Performance Metrics in the GUI:

https://earth.bsc.es/autosubmitapp/experiment/a2s5

It will be showed tomorrow in the CES meeting.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 4, 2020, 11:57

It definitely requires much more work. Suggestions are welcome.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 4, 2020, 12:08

@wuruchi , regarding your message about the rest of the metrics:

* The energy consumption is available through Slurm. This is related with #484 and it could be done at the same time, providing also that info in the job logs.

* For the rest of the metrics, I think you can have a look at the Auto-EC-Earth Perf. Metrics template:

https://earth.bsc.es/gitlab/es/auto-ecearth3/-/blob/trunk/templates/common/performance_metrics.tmpl.sh

Some metrics as the RSYPD can be ported to AS.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 4, 2020, 12:11

Taking a look.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 4, 2020, 13:13

@mcastril @pablohe 

From what I am seeing, I can infer that not all %SIM%.out files will include the performance metrics (for example energy consumption, ASYPD, SYPD, etc.) in the same format or might include only some of them. Is that correct? I guess it depends on the model the user is running his experiment on.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 4, 2020, 13:39

Regarding energy consumption, ASYPD, SYPD... we can include them in all SIM jobs because this is kind of an standard. Maybe for MONARCH we can change to a SDPD, this can be triggered when the CHUNKSIZE is day instead of month or year.

There are other metrics that are more model specific and that we will need to create an interface between the auto-model and autosubmit.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 4, 2020, 13:41

Yep. I was looking at the auto-ecearth code and comparing it to experiments results and it seems pretty hard to get the values from the .out files in a reliable manner.

I guess we can start by adding a slurm request in Autosubmit that gets the energy consumption of the job independent of the model used.

**Comment by kinow** [comment]
In GitLab by @ macosta on Jun 4, 2020, 13:52

@wuruchi for the other metrics, please take a look but actually we should discuss about the implementation. I think we can talk any moment next week once you have clear these metrics, then we can discuss about the implementation.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 4, 2020, 13:56

Agree, @ macosta. I think the focus is gravitating more towards getting more information for Autosubmit. The actual implementation will be hold until the previous point is defined.

**Comment by kinow** [comment]
In GitLab by @pablohe on Jun 4, 2020, 17:10

Hi, 
in this https://earth.bsc.es/autosubmitapp/experiment/t01q experiment
that I run every week more or less, I see 

```
Parallelization: 768
SYPD: 14400
ASYPD: 14400
CHSY: 0
```

It is ok that the GUI show those numbers: 14400

and 

Parallelization is `SIM` number of cores

Is it any reason to change the named ? It is only a question

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jun 5, 2020, 08:22

Hello @pablohe 

The 14400 bug happens when there are no SIM jobs. I will fix it in the next iteration.

>Parallelization is `SIM` number of cores

Yes, it is. We adopted that name because that is how it is named in the provided documentation.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jul 13, 2020, 13:16

mentioned in commit c138dbdf0368504184c34b3d4c9cd7f9adf76809

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 14, 2020, 15:44

We have been discussing briefly about the next steps. Summarizing:

1. Most of the metrics in https://earth.bsc.es/gitlab/es/autosubmit/issues/524#note_80848 and the `Memory bloat` can be directly calculated in Autosubmit.

2. The remaining ones, which are `Coupling Cost`, `Data output cost`, `Data intensity`, that are more model dependent and more difficult to measure, we can rely in the current template and provide a way to interface this with Autosubmit, as we discussed previously.

For the metrics in 1) we have 2 main doubts:

* Which parameter we should look for in `sacct` to get the memory bloat?
  * Looking an old conversation between @ macosta and David Vicente, he recommend to use `AveRSS` or `MaxRSS`.
  * An alternative approach, not using Slurm, is to use the `time` command, but again this should be on the Auto-Model side. I think it is better to use Autosubmit if we can.
  * The current implementation seems to rely in the epilog from the SIM job. This is written at the end of the job logs in Nord3 (I guess it uses MaxRSS), but not in MN4. I guess you only get this information in Nord3 @pablohe right?

* The second doubt is about ASYPD and RSYPD. I always (or lately only because I just forgot) thought that ASYPD was RUN+QUEUE, and RSYPD was counting the full critical path. However, looking at the doc, ASYPD should also count stops and interruptions. RSYPD is not detailed in the documentation https://docs.google.com/document/d/12yWDwXsohf4G4MPeP6e3Eil4ZL-YeIN71dBcoWRliEg/edit and the implementation in the template doesn't help because it is just the same: https://earth.bsc.es/gitlab/es/auto-ecearth3/-/blob/trunk/templates/common/performance_metrics.tmpl.sh#L158 (I think we did this manually for CMIP6 @ macosta ?).

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 14, 2020, 15:45

In the EC-Earth forum I found a deeper explanation of ASYPD/RSYPD by @ macosta :

https://dev.ec-earth.org/issues/532

> We are dividing this in two metrics:
> - ASYPD is the same that Philippe is suggesting, including queue and run time for the EC-Earth run script, but we are also including queue and
> run time for post processing, so we understand that it is a value indicating the complete process to generate outputs of 1 year. We are using Autosubmit logs to take an accumulative total time of Run jobs and the last Post, assuming that the others are running in parallel. Philippe approach should be feasible for different platforms, but some specific development would needed for each one. Another question appears here for our BSC approach, we are not including the resources of postprocessing for metrics as CHPSY or parallelization...
> - RSYPD is measuring the start date of the submission of the first job and the final date of the last job. This means that we are including the
> time needed when some job fails and the experiment stops, waiting for someone to rerun a particular chunk for example (quota problems, nodes failing...) you can surprise about how different are ASYPD and RSYPD. We are using Autosubmit logs.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 14, 2020, 15:51

So some ideas here:

* For ASYPD, if we counted all the CHUNK jobs depending on SIM, it should be very easy to do and generalizable. Of course we can count only the POST, but then it will not be usable by other models. Given that EC-Earth3 cmorizes the data in two independent jobs (CMORATM and CMOROCE) before the POST, I think that both CMOR jobs should be count as POST. In that case we could count all the remote jobs after SIM (CLEAN and TRANSFER run in the transfer nodes).

* The RSYPD is not much complex after we define the ASYPD. What do we understand as "final date"? If it is the date of the POST, we should do the same as what we decide to do in the ASYPD.

**Comment by kinow** [comment]
In GitLab by @ macosta on Jul 14, 2020, 15:59

Thank you Miguel for the research work! I will start for the ASYPD/RSYPD discussion.

Here we are improving the community definition so we have not relay on other papers. Our agreement should to include in ASYPD those jobs needed only to run and postprocess (prepare) the outputs. It could be CMOR jobs too but I would not introduce CLEAN and transfer since it is more machine dependent.

RSYPD as you say it is easier once we define ASYPD, which should include the complete workflow and the interruptions/variability of a machine.

Other thing that we could do it is to include an additional one. I mean, ASYPD could be CMOR+RUN+POST (only jobs to produce outputs). RSYPD the complete workflow including clean, transfers and other stuff. Finally, we could include something like FSYPD (Final SYPD), which should measure the time from the beginning to the end to include interruptions and variability.

What do you think?

**Comment by kinow** [comment]
In GitLab by @ macosta on Jul 14, 2020, 16:03

COupling cost should be included by Sergi implementation (LUCIA) and probably provided by the automodel output as Data Ouput cost and Data intensity as you say.

For memory bloat, I agree that we could use MaxRSS if it is possible.

Finally I have an additional question, are you taking into account those metrics that we included in the STATS job and where they will be calculated and saved? For example, the execution time for each job separately, Max/MIN/AVG time for jobs as SIM or POST...

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 14, 2020, 16:17

> Here we are improving the community definition so we have not relay on other papers. Our agreement should to include in ASYPD those jobs needed only to run and postprocess (prepare) the outputs. It could be CMOR jobs too but I would not introduce CLEAN and transfer since it is more machine dependent. 

Perfect. As we run CLEAN and TRANSFER in local (or data transfer) nodes, they should not be counted. 

> Other thing that we could do it is to include an additional one. I mean, ASYPD could be CMOR+RUN+POST (only jobs to produce outputs). RSYPD the complete workflow including clean, transfers and other stuff. Finally, we could include something like FSYPD (Final SYPD), which should measure the time from the beginning to the end to include interruptions and variability. 

I think it is important to count the transfers in some way, because in the end is part of the workflow and users normally work with the data locally. I don't know if this is for all the institutions (maybe some use the data remotely?).

What I find more difficult is to automatically identify what are post-processing and data-transfer jobs. We can consider that anything running in the cluster after the SIM is POST, as we said previously. But the rest of the jobs can combine clean, transfer, archive, diagnostics... We have to give it a think. 

For Auto-MONARCH, in the DA workflow they are now running REDUCE + CALC_STATS remotely (that would be the POST) and then they run ARCHIVE (that would be the TRANSFER). But I think that sometimes REDUCE or CALC_STATS are run offline.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 14, 2020, 16:19

@gmontanepinto is the one that knows better about Auto-MONARCH and how we could converge.

If we don't find a way to do it automatically, we could maybe create an optional `parameter` in the jobs.conf , to tag any section (job) as SIM, POST or TRANSFER. In that case they would not be forced to call SIM to the simulation either.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 14, 2020, 16:21

> Finally I have an additional question, are you taking into account those metrics that we included in the STATS job and where they will be calculated and saved? For example, the execution time for each job separately, Max/MIN/AVG time for jobs as SIM or POST... 

@wuruchi created a distributed database that will start to record that values for every job in the workflow. So yes.

https://earth.bsc.es/gitlab/es/autosubmit/issues/553

**Comment by kinow** [comment]
In GitLab by @gmontanepinto on Jul 14, 2020, 18:53

In auto-monarch the REDUCE job is always run, and the CALC_STATS is only run in the DA case. 

>  We can consider that anything running in the cluster after the SIM is POST

Both jobs can be run either in MN4 or Nord3 (as well as the POST job in auto-ecearth I guess), so I think that looking at the platform is not enough as it can be different than the one where the SIM runs. Right now I can't think of an easy common solution to know which are the POST jobs without adding more info in the jobs definition.

**Comment by kinow** [comment]
In GitLab by @pablohe on Jul 15, 2020, 07:56

> Thanks @gmontanepinto . In fact also Auto-EC-Earth has the capability to run CMOR jobs in Nord3, although we usually run them in the same platform as the SIM, to avoid problems in PRACE/RES workflows. 

CMORIZATION doesn't work in Nord3, @ aamaral could you please confirm this?


> The current implementation seems to rely in the epilog from the SIM job. This is written at the end of the job logs in Nord3 (I guess it uses MaxRSS), but not in MN4. I guess you only get this information in Nord3 @pablohe right?

It is not referenced in the code about `MaxRSS`

**Comment by kinow** [comment]
In GitLab by @ aamaral on Jul 15, 2020, 11:18

Hi @pablohe 

From what I remember it could work. I saw that I modified the configure.sh file on platforms/nord3, but to be honest I can't remember if it was tested. I usually ran the cmorisation itself on MN4.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 28, 2020, 11:16

Thanks @wuruchi . It's true that usually each chunk has its post-processing and transfer jobs, so at the end SIMs may be 20% or 25% of the jobs. This raises the interesting question of including the energy consumption of the post-processing or not. This are usually serial jobs (using one full node) but usually they take some time, and in SR experiments where we run SIMs with 10-15 nodes maybe it is not so negligible.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jul 29, 2020, 13:21

Hello @mcastril @pablohe @ macosta @gmontanepinto 

I have added `JPSY` to the current implementation of the Autosubmit API and the GUI. The Autosubmit API [Performance Metrics API call](https://earth.bsc.es/gitlab/wuruchi/autosubmitreact/-/wikis/Performance-Metrics) has been updated to reflect these changes.

Also, the GUI has been updated to show this metric and the energy value from the job historical database: https://earth.bsc.es/autosubmitapp/experiment/t023

You might need to refresh the web page to see the changes.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Aug 14, 2020, 18:00

Important: The feature is currently showing inaccurate data. There was an undetected issue with the data collection function that is now being addressed.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 19, 2021, 11:48

@wuruchi you can review the conversation that we had about ASYPD / RSYPD:

* ASYPD should count SIM + POSTPROCESSING running & queueing time for a given chunk. The metric is assigned to the SIM of the corresponding chunk.

* RSYPD should count the wallclock time (including workflow interruptions) between the initialization of the SIM and the finalization of the chunk (so including CLEAN & TRANSFER jobs). The metric is also assigned to that SIM.

So we are aggregating running and queuing times in ASYPD, while in RSYPD we are just subtracting timestamps.

In order to distinguish SIM and POST jobs we need some meta-information that we can get from the TASK_TYPE parameter:
* For SIMs I think we can account any job called SIM or with TASK_TYPE = SIM*
* For POST we can add any job called POST* or having TASK_TYPE = POST* or CMOR* (it's possible that we also need to use the TASK_TYPE argument to run different CMOR jobs using the same template).

What do you think @gmontanepinto @ jberlin . Could we add this TASK_TYPE for the SIM, POST, CMOR, REDUCE, CALC_STATS, etc jobs of the different testing suites?

PS: Notice that in the current implementation of Auto-EC-Earth metrics there's no distinction between ASYPD and RSYPD, we postponed this development.

@ macosta

**Comment by kinow** [comment]
In GitLab by @gmontanepinto on Jan 19, 2021, 17:46

Fine for me!

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 19, 2021, 21:01

Hello @mcastril 

From `Balaji et al.` we have that:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/7df568c56328bce92e57f6048c37fc53/image.png)

The latest implementation of `ASYPD` considers $`t_0`$ as the `submit time` of the first job in the experiment, $`t_N`$ as the `finish time` of the latest `SIM` job in the experiment, and $`N`$ is the number of simulated years in total until the latest `SIM` job. 

> ASYPD should count SIM + POSTPROCESSING running & queueing time for a given chunk. The metric is assigned to the SIM of the corresponding chunk.

According to this description, I understand that we should not follow the `t_0` definition for the metric in the paper and consider only the times for the `SIM` and `POSTPROCESSING` jobs.

> RSYPD should count the wallclock time (including workflow interruptions) between the initialization of the SIM and the finalization of the chunk (so including CLEAN & TRANSFER jobs). The metric is also assigned to that SIM.

I understand that `RSYPD` should start counting time from the `start time` of the `SIM` job until the `finish time` of `CLEAN` or `TRANSFER` (whichever finishes latest) of the same chunk.

> For SIMs I think we can account any job called SIM or with TASK_TYPE = SIM\*

It seems that this `TASK_TYPE` variable is not defined in the `jobs_[expid].conf`. Where can we find it?

So far, we can only identify `SIM` type jobs by testing for the inclusion of the string `SIM` in the job's name.

**Comment by kinow** [comment]
In GitLab by @ macosta on Jan 20, 2021, 10:04

Hello everyone,

It is as @mcastril is commenting. Sorry for the confusion, for some metrics we decided to go beyond the paper. So the idea is to save (at least quantify) every SIMi, POSTi... With this approach we calculate ASYPD as Miguel was commenting but we can also print the average and accumulative time of SIM or POST jobs, or even create new metics in the future if it is needed.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 20, 2021, 10:27

Thanks @ macosta . We save all the data and metrics are calculated on the fly, so this gives us the possibility to change things. By now I think @wuruchi can implement the calculus as in my last message and then we will have a review.

You are right @gmontanepinto . I saw Etienne was using it with the value `SPINUP_FORCING` and I didn't realize that was the actual name of the job. 

What do you think we should do? We have two options I think:

- Either creating a new variable, which would force to have the correct values filled for the metrics to work.
- Or either hardcoding some usual names in the GUI TASKTYPE=(CMOR*, POST*, REDUCE*...)

I think none of both approaches is ideal. But it maybe is less problematic to have them hardcoded as it is changing one line vs defining one parameter in lots of configurations.

**Comment by kinow** [comment]
In GitLab by @ jberlin on Jan 20, 2021, 11:39

Hi @mcastril , fine for me ! as mentioned by @gmontanepinto , the TASKTYPE is already there. I used it in the refactor of the CMOR templates, can´t we use the same one ?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 20, 2021, 13:23

Hello @mcastril 

> So, to define for once the metric, would it be ok to have:
```
ASYPD: Sum(Sum(Queuing(SIMi), Running(SIMi)) for i=1,N , Sum(Queuing(POSTi) + (Running(POSTi) i=N)
RSYPD: Substract(EndTime(TRANSFERi) i=N , StartTime(SIMi) i=1)
```

For the record, we define:

* $`Status() = `$ Function that returns the status of a job as a string.
* $`Run() = `$ Function that returns the running time of a `COMPLETED` job in seconds.
* $`Queue() = `$ Function that returns the queueing time of a `COMPLETED` job in seconds.
* $`Start() = `$ Returns the start time of a job (not the **submitted** time) in seconds.
* $`Finish() = `$ Returns the finish time of a job in seconds.
* $`Years() = `$ Function that returns the number of simulated years of a `SIM` (or identified as a such) job. Calculated from the experiment's `chunk size` and `chunk unit`.
* $`N = `$ Number of chunks $`j`$ such that $`Status(SIM_j) = Status(POST_j) = COMPLETED`$.
* $`M = `$ Number of chunks $`k`$ such that $`Status(SIM_k) = Status(TRANSFER_k) = COMPLETED`$.

$`ASYPD = \frac{(\sum_{i}^{N} Years(SIM_i))*86400}{\sum_{i}^{N} Queue(SIM_i) + Run(SIM_i) + Queue(POST_i) + Run(POST_i)}`$

$`RSYPD = \frac{(\sum_{i}^{M} Years(SIM_i))*86400}{\sum_{i}^{M} Finish(TRANSFER_i) - Start(SIM_i)}`$ 

Where the units are $`\frac{years*seconds/day}{seconds} = years/day`$

### Corrected versions:


$`ASYPD = \frac{(\sum_{i}^{N} Years(SIM_i))*86400}{(\sum_{i}^{N} Queue(SIM_i) + Run(SIM_i))+ Queue(POST_n) + Run(POST_n)}`$

$`RSYPD = \frac{(\sum_{i}^{M} Years(SIM_i))*86400}{Finish(TRANSFER_m) - Start(SIM_1)}`$

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 20, 2021, 19:15

Hello @mcastril 

Thank you for the corrections. Here are the formulas that will be implemented:

$`ASYPD = \frac{(\sum_{i}^{N} Years(SIM_i))*86400}{(\sum_{i}^{N} Queue(SIM_i) + Run(SIM_i))+ Queue(POST_n) + Run(POST_n)}`$

$`RSYPD = \frac{(\sum_{i}^{M} Years(SIM_i))*86400}{Finish(TRANSFER_m) - Start(SIM_1)}`$

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 21, 2021, 13:03

mentioned in commit 9094128b7495df2889d6a39ed6bdc26b4b4bdeff

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 22, 2021, 12:43

Hello @mcastril 

> However, there must be something with the RSYPD. It never can be higher than ASYPD. Is it because `a33f` does not have a `TRANSFER`? But I don't see the warning.

Warnings are shown in the warning section:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/ca9e30fd27154e25259278d29ab7cb77/image.png)

Now I notice that the warnings says "`ASYPD` cannot be computed." It should say `RSYPD` instead. I'm going to fix it.

> Nevertheless, I forgot to add that currently most of the EC-Earth workflows do not have a `TRANSFER` job because we move the data directly on `CLEAN`. Could you use `CLEAN` instead of `TRANSFER` for `RSYPD` in case `TRANSFER` is missing?

Sure, I am going to add it.

> I have another question. In experiments having multiple start dates or members, which POST are you taking into account for `ASYPD`? I think the safest approach would be the average of all last-chunk POSTs.

I am considering the latest POST, and then I only consider the SIM jobs with a finish time lesser than the finish time of that POST. An average would not be a problem.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 22, 2021, 12:51

Good, thanks!

Wrt to the Warning, I was talking about `a33f` does do not have any `TRANSFER`. It has `TRANSFER_MEMBER`, on the other side, that is not the same, but it could pass the filter maybe as we were talking about wildcards above (to make it more "compatible").

So I think we should better rather looking for "CLEAN" and "TRANSFER" to avoid problems with CLEAN_MEMBER and TRANSFER_MEMBER, that do not move data.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 25, 2021, 10:23

mentioned in commit 8fc799adebc38a3c571b13c15a67d2cb4907c7dc

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 25, 2021, 14:52

Thanks for the correct specification @wuruchi.

I have one only amendment:

In `ASYPD`, the N for SIM is different than the N for POST. In SIM we iterate through chunks, and in POST we iterate through startdates / members.

In fact, this is not anything particular of the `ASYPD` and it should be common to all metrics (SYPD, RSYPD too). In any experiment having multiple members / startdates, all these metrics should be `averages` over those levels (member / startdate). So you maybe can leave just N in POST (without the sum) and we will just specify the metric for a given member. Then we remind that all are averages at the experiment level. Sorry if it was not so clear in my previous message.

And just another comment about the dynamic update of the metrics:

If a user wants to query the `ASYPD` or `RSYPD` of a running experiment (same as they can do with SYPD), I think we should provide some data, even if the experiment is not finished. So I think that you could simply provide the metric for the last completed chunk, i.e. chunk run until POST for ASYPD, or chunk run until CLEAN/TRANS for RSYPD.

**Comment by kinow** [comment]
In GitLab by @wuruchi on May 31, 2021, 12:26

As I see it, the objectives of this issue have been reached. I will change it to documentation and close it. New developments on the topic of Performance Metrics would require a new issue that specifies those.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 6, 2024, 10:55

mentioned in issue autosubmit-api#75
